Convert MySQL fields to 4 byte unicode

dimaursu opened the pull request:
This will make Diaspora\* running on postgres and on mysql behave consistently: in the postgres version, you can use 4-byte UTF8, or full UTF8. In MySQL `utf8` == the 3 byte subset of UTF8, while `utf8mb4` is the true, full UTF8 encoding.

jhass commented on db/schema.rb:
```diff
@@ -429,11 +433,11 @@
+    t.string   "item_type",                  null: false
     t.integer  "user_id",                    null: false
   end
 
```
No actual schema changes in the migration, please don't commit schema.rb changes.

jhass commented on db/migrate/20150106050733_set_mysql_to_unicode_mb4.rb:
```diff
@@ -0,0 +1,38 @@
+  }
+
+  def self.up
+    if ENV['DB'] == 'mysql'
```
I plan to eventually get rid of the DB environment variable, I think you should be able to query the adapter through AR::Base.config or something like that.

jhass replied:
Actually I just remembered we already have https://github.com/diaspora/diaspora/blob/develop/lib/configuration_methods.rb#L121 which can be called through `AppConfig.postgres?`. Don't need that often luckily!

jhass commented:
I guess you want to add to the big fat changelog entries here, so podmins notice to update their database.yml.

jhass commented on db/schema.rb:
```diff
@@ -11,7 +11,10 @@
 # It's strongly recommended that you check this file into your version control system.
 
-ActiveRecord::Schema.define(version: 20141216213423) do
+ActiveRecord::Schema.define(version: 20150106050733) do
```
That is, the version change should be committed of course.

jhass commented on Gemfile:
```diff
@@ -57,7 +57,7 @@ gem 'sass-rails',     '4.0.4'
 # Database
 
-ENV['DB'] ||= 'mysql'
+ENV['DB'] ||= 'postgres'
```
As much as I'd wish for, but nope...

jhass commented on Gemfile:
```diff
@@ -88,6 +88,7 @@ gem 'rails-assets-jquery',   '1.11.1' # Should be kept in sync with jquery-rails
 gem 'js_image_paths',        '0.0.1'
 gem 'js-routes',             '0.9.9'
 gem 'rails-assets-punycode', '1.3.2'
+gem 'rails-assets-twemoji'
```
Let's keep that change separate, since I disagree with it while agreeing about changing the database encoding.

goobertron commented:
If this is merged, please can we also have a means of blocking these emoji from view?

dimaursu committed:
Add warning about encoding change

Add warning about MySQL collation

Fix database index length

This allows new databases to be created with utf8mb4, on MySQL. The maximum
column size is 767 bytes. Each character is 4 bytes long -> 767 / 4 = 191
characters for the column.

jhass commented:
@goobertron it's actually not graphic emojis for this one and actually just badly argued through the unicode emojis, so it's rather unreleated to them, they're just a side effect. There's strong technical reasons to do this.

goobertron commented:
Oh right, I was basing my comment on the opening post, which includes graphics. I guess that's Github's fault. I'd still like an opt-out from seeing any emoji, however they are displayed.

jhass commented:
For this one, just remove any fonts that would display them... :P

dimaursu committed:
Refactor & DRY encoding migration

Dynamic row for MySQL

Set larger column index sizes

dimaursu commented on db/migrate/20150106050733_set_mysql_to_unicode_mb4.rb:
```diff
@@ -0,0 +1,37 @@
+  end
+
+  def self.up
+    change_encoding('utf8mb4') if AppConfig.mysql?
```
@jhass ![I'ts beautiful!](http://www.quickmeme.com/img/30/30bea914dbe40c224d53e671e903efb8cd6bafe2b599f8fd341ed8136af75f23.jpg)

jhass commented on db/migrate/20150106050733_set_mysql_to_unicode_mb4.rb:
```diff
@@ -0,0 +1,37 @@
+    ActiveRecord::Base.connection.columns(table).each do |column|
+      # build a hash with all the columns that contain text
+      if (column.type == :string) || (column.type == :text)
+        UTF8_PAIRS[table] = { 'name' => column.name, 'type' => column.sql_type }
```
I'd use symbol keys here :)

dimaursu commented:
@jhass `ROW_FORMAT=DYNAMIC`, in combination with InnoDB engine, and those 3 parameters above (now set in the migration, no need to manually touch configs), increases the maximum size of the index columns from 767 bytes to 3072 (something like that); that gives us 768 chars for the indexes, in the worst case scenario (a whole index row is made out of emojis would be a good example). MySQL only considers this scenario, that's why it spits out errors.

jhass commented:
Okay, I see now how that transforms existing databases. How do we get to the same state when doing `rake db:create db:schema:load`?

dimaursu commented:
No clue about that. But if you give me an idea or some pointers, I will gladly add that too.

dimaursu commented on db/migrate/20150106050733_set_mysql_to_unicode_mb4.rb:
```diff
@@ -0,0 +1,43 @@
+    ActiveRecord::Base.connection.columns(table).each do |column|
+      # build a hash with all the columns that contain text
+      if (column.type == :string) || (column.type == :text)
+        UTF8_PAIRS[table] = { :name => column.name, :type => column.sql_type }
```
This is not good. We should have the key the column. This is a serious error from my side, sorry.

dimaursu committed:
Shorten indexes

Fix merge conflict

jhass commented on db/schema.rb:
```diff
@@ -31,19 +31,19 @@
+    t.integer  "aspect_id",      null: false
+    t.datetime "created_at",     null: false
+    t.datetime "updated_at",     null: false
+    t.string   "shareable_type"
```
Where did the default went?

jhass replied:
This is still an issue.

jhass commented on db/schema.rb:
```diff
@@ -31,19 +31,19 @@
 
   create_table "aspects", force: true do |t|
-    t.string   "name",                             null: false
+    t.string   "name"
```
Where did the not null constraint went?

jhass replied:
This too is still an issue.

jhass commented on db/schema.rb:
```diff
@@ -61,47 +61,47 @@
-    t.integer "user_id",                  null: false
-    t.string  "jid",                      null: false
+    t.integer "user_id",                   null: false
+    t.string  "jid",          limit: 3071
```
Left over/typo?

jhass commented on db/schema.rb:
```diff
@@ -61,47 +61,47 @@
   end
 
-  add_index "chat_contacts", ["user_id", "jid"], name: "index_chat_contacts_on_user_id_and_jid", unique: true, using: :btree
+  add_index "chat_contacts", ["user_id", "jid"], name: "index_chat_contacts_on_user_id_and_jid", length: {"user_id"=>nil, "jid"=>190}, using: :btree
```
uniqueness constraint went missing.

jhass commented on db/schema.rb:
```diff
@@ -61,47 +61,47 @@
+    t.datetime "created_at",                                     null: false
+    t.datetime "updated_at",                                     null: false
+    t.integer  "likes_count",                        default: 0, null: false
+    t.string   "commentable_type",        limit: 60
```
default, not null

jhass commented on db/schema.rb:
```diff
@@ -366,7 +366,7 @@
     t.string   "actor_url"
     t.string   "objectId"
-    t.string   "root_guid"
+    t.string   "root_guid",             limit: 64
```
New limit not present in migration.

jhass replied:
Still an issue.

jhass commented on db/schema.rb:
```diff
@@ -376,18 +376,18 @@
     t.boolean  "favorite",                         default: false
     t.string   "facebook_id"
-    t.string   "tweet_id"
+    t.string   "tweet_id",              limit: 64
```
New limit

jhass commented on db/schema.rb:
```diff
@@ -457,16 +457,16 @@
+    t.datetime "updated_at",                                null: false
+    t.boolean  "hidden",                    default: false, null: false
+    t.integer  "contact_id",                                null: false
+    t.string   "shareable_type", limit: 60
```
default, not null

jhass commented on db/schema.rb:
```diff
@@ -529,8 +529,8 @@
-    t.string   "email",                                          default: "",    null: false
-    t.string   "encrypted_password",                             default: "",    null: false
+    t.string   "email"
+    t.string   "encrypted_password"
```
default, not null

jhass commented on db/schema.rb:
```diff
@@ -561,10 +561,10 @@
-  add_index "users", ["email"], name: "index_users_on_email", using: :btree
-  add_index "users", ["invitation_service", "invitation_identifier"], name: "index_users_on_invitation_service_and_invitation_identifier", unique: true, using: :btree
+  add_index "users", ["email"], name: "index_users_on_email", length: {"email"=>191}, using: :btree
+  add_index "users", ["invitation_service", "invitation_identifier"], name: "index_users_on_invitation_service_and_invitation_identifier", unique: true, length: {"invitation_service"=>64, "invitation_identifier"=>nil}, using: :btree
```
The nil is intentional?

Flaburgan commented on Changelog.md:
```diff
@@ -32,6 +32,10 @@ series and run our comprehensive test suite against it.
 
+## Change in database.yml
+For MySQL databases, replace `charset: utf8` with `encoding: utf8mb4`. This is enables full UTF8 support (4bytes characters), including standard emoji characters. See `database.yml.example` for reference.
+Also, do not forget to remove `collation: utf8_bin`. It will choose a compatible one automatically.
```
This is not precise enough imo. Where is that setting? Which file, which line? If you really need a podmin action, you should explain precisely what he should do.

dimaursu commented on Changelog.md:
```diff
@@ -32,6 +32,10 @@ series and run our comprehensive test suite against it.
 
+## Change in database.yml
+For MySQL databases, replace `charset: utf8` with `encoding: utf8mb4` in the file `config/database.yml`. This is enables full UTF8 support (4bytes characters), including standard emoji characters. See `database.yml.example` for reference.
+Also, do not forget to remove `collation: utf8_bin`. It will choose a compatible one automatically.
```
With `database.yml.example` at hand, I guess any podmin will figure this out.

jhass commented:
https://travis-ci.org/diaspora/diaspora/jobs/46997679#L357

denschub commented:
As promised, ran the migration on a clone of my production database.

Size on disk before: 43062 MB
Size on disk after: 43167 MB

Full migration: 304.7380s, full log below:

```
== 20150106050733 SetMysqlToUnicodeMb4: migrating =============================
-- remove_index("aspect_visibilities", {:name=>"shareable_and_aspect_id"})
   -> 0.0574s
-- add_index("aspect_visibilities", ["shareable_id", "shareable_type", "aspect_id"], {:name=>"shareable_and_aspect_id", :length=>{"shareable_type"=>189}, :using=>:btree})
   -> 16.0878s
-- remove_index("aspect_visibilities", {:name=>"index_aspect_visibilities_on_shareable_id_and_shareable_type"})
   -> 0.0500s
-- add_index("aspect_visibilities", ["shareable_id", "shareable_type"], {:name=>"index_aspect_visibilities_on_shareable_id_and_shareable_type", :length=>{"shareable_type"=>190}, :using=>:btree})
   -> 9.1848s
-- remove_index("chat_contacts", {:name=>"index_chat_contacts_on_user_id_and_jid"})
   -> 0.1278s
-- add_index("chat_contacts", ["user_id", "jid"], {:name=>"index_chat_contacts_on_user_id_and_jid", :length=>{"jid"=>190}, :using=>:btree, :unique=>true})
   -> 0.1719s
-- remove_index("comments", {:name=>"index_comments_on_guid"})
   -> 0.2913s
-- add_index("comments", ["guid"], {:name=>"index_comments_on_guid", :length=>{"guid"=>191}, :using=>:btree, :unique=>true})
   -> 33.2449s
-- remove_index("likes", {:name=>"index_likes_on_guid"})
   -> 0.3831s
-- add_index("likes", ["guid"], {:name=>"index_likes_on_guid", :length=>{"guid"=>191}, :using=>:btree, :unique=>true})
   -> 70.5352s
-- remove_index("o_embed_caches", {:name=>"index_o_embed_caches_on_url"})
   -> 0.0831s
-- add_index("o_embed_caches", ["url"], {:name=>"index_o_embed_caches_on_url", :length=>{"url"=>191}, :using=>:btree})
   -> 2.0182s
-- remove_index("participations", {:name=>"index_participations_on_guid"})
   -> 0.1053s
-- add_index("participations", ["guid"], {:name=>"index_participations_on_guid", :length=>{"guid"=>191}, :using=>:btree})
   -> 96.0637s
-- remove_index("people", {:name=>"index_people_on_diaspora_handle"})
   -> 0.0838s
-- add_index("people", ["diaspora_handle"], {:name=>"index_people_on_diaspora_handle", :unique=>true, :length=>{"diaspora_handle"=>191}})
   -> 0.9094s
-- remove_index("people", {:name=>"index_people_on_guid"})
   -> 0.0497s
-- add_index("people", ["guid"], {:name=>"index_people_on_guid", :length=>{"guid"=>191}, :using=>:btree, :unique=>true})
   -> 0.3075s
-- remove_index("photos", {:name=>"index_photos_on_status_message_guid"})
   -> 0.0415s
-- add_index("photos", ["status_message_guid"], {:name=>"index_photos_on_status_message_guid", :length=>{"status_message_guid"=>191}, :using=>:btree})
   -> 2.8110s
-- remove_index("posts", {:name=>"index_posts_on_guid"})
   -> 0.0748s
-- add_index("posts", ["guid"], {:name=>"index_posts_on_guid", :length=>{"guid"=>191}, :using=>:btree, :unique=>true})
   -> 26.3498s
-- remove_index("posts", {:name=>"index_posts_on_status_message_guid_and_pending"})
   -> 0.1532s
-- add_index("posts", ["status_message_guid", "pending"], {:name=>"index_posts_on_status_message_guid_and_pending", :length=>{"status_message_guid"=>190}, :using=>:btree})
   -> 6.0173s
-- remove_index("posts", {:name=>"index_posts_on_status_message_guid"})
   -> 0.0775s
-- add_index("posts", ["status_message_guid"], {:name=>"index_posts_on_status_message_guid", :length=>{"status_message_guid"=>191}, :using=>:btree})
   -> 6.4922s
-- remove_index("rails_admin_histories", {:name=>"index_rails_admin_histories"})
   -> 0.0716s
-- add_index("rails_admin_histories", ["item", "table", "month", "year"], {:name=>"index_rails_admin_histories", :length=>{"table"=>188}, :using=>:btree})
   -> 0.1415s
-- remove_index("schema_migrations", {:name=>"unique_schema_migrations"})
   -> 0.7815s
-- add_index("schema_migrations", ["version"], {:name=>"unique_schema_migrations", :length=>{"version"=>191}, :using=>:btree})
   -> 0.2091s
-- remove_index("services", {:name=>"index_services_on_type_and_uid"})
   -> 0.0873s
-- add_index("services", ["type", "uid"], {:name=>"index_services_on_type_and_uid", :length=>{"type"=>64, "uid"=>127}, :using=>:btree})
   -> 0.2988s
-- remove_index("taggings", {:name=>"index_taggings_on_taggable_id_and_taggable_type_and_context"})
   -> 0.0946s
-- add_index("taggings", ["taggable_id", "taggable_type", "context"], {:name=>"index_taggings_on_taggable_id_and_taggable_type_and_context", :length=>{"taggable_type"=>95, "context"=>95}, :using=>:btree})
   -> 23.9894s
-- remove_index("tags", {:name=>"index_tags_on_name"})
   -> 0.0702s
-- add_index("tags", ["name"], {:name=>"index_tags_on_name", :length=>{"name"=>191}, :using=>:btree, :unique=>true})
   -> 1.5240s
-- remove_index("users", {:name=>"index_users_on_invitation_service_and_invitation_identifier"})
   -> 0.1581s
-- add_index("users", ["invitation_service", "invitation_identifier"], {:name=>"index_users_on_invitation_service_and_invitation_identifier", :length=>{"invitation_service"=>64, "invitation_identifier"=>127}, :using=>:btree, :unique=>true})
   -> 0.5600s
-- remove_index("users", {:name=>"index_users_on_username"})
   -> 0.1131s
-- add_index("users", ["username"], {:name=>"index_users_on_username", :length=>{"username"=>191}, :using=>:btree, :unique=>true})
   -> 0.2943s
-- remove_index("users", {:name=>"index_users_on_email"})
   -> 0.1425s
-- add_index("users", ["email"], {:name=>"index_users_on_email", :length=>{"email"=>191}, :using=>:btree})
   -> 0.1858s
-- remove_index("notifications", {:name=>"index_notifications_on_target_type_and_target_id"})
   -> 0.1081s
-- add_index("notifications", ["target_type", "target_id"], {:name=>"index_notifications_on_target_type_and_target_id", :length=>{"target_type"=>190}, :using=>:btree})
   -> 4.1315s
== 20150106050733 SetMysqlToUnicodeMb4: migrated (304.7380s) ==================
```

:+1:

dimaursu committed:
Change the adapter encoding to utf8mb4

jhass commented on db/schema.rb:
```diff
@@ -155,7 +155,7 @@
     t.string   "identifier"
     t.boolean  "admin",        default: false
-    t.string   "language",     default: "en"
+    t.string   "language"
```
This one is making the tests fail, which doesn't mean the other ones shouldn't be fixed too.

jhass commented:
Alright, let's risk this :P

Thank you!

Flaburgan commented:
I run that on diaspora-fr and have http://paste.mrzyx.de/pxfy0zy2m
